From 7798ae25574425271305fffe85de77bec8df03f1 Mon Sep 17 00:00:00 2001
From: Kevin Backhouse <kev@semmle.com>
Date: Wed, 15 May 2019 10:12:02 +0100
Subject: [PATCH] Throw an exception if the data location is invalid. (#842)

[Note - removed binary image from patch]
---
 src/crwimage_int.cpp                    |   7 +++----
 src/crwimage_int.hpp                    |   1 -
 test/data/issue_841_poc.crw             | Bin 0 -> 10078 bytes
 tests/bugfixes/github/test_issue_841.py |  22 ++++++++++++++++++++++
 4 files changed, 25 insertions(+), 5 deletions(-)
 create mode 100644 test/data/issue_841_poc.crw
 create mode 100644 tests/bugfixes/github/test_issue_841.py

diff --git a/src/crwimage_int.cpp b/src/crwimage_int.cpp
index 4080c078..58f1f467 100644
--- a/src/crwimage_int.cpp
+++ b/src/crwimage_int.cpp
@@ -570,12 +570,11 @@ namespace Exiv2 {
 
     DataLocId CiffComponent::dataLocation(uint16_t tag)
     {
-        DataLocId di = invalidDataLocId;
         switch (tag & 0xc000) {
-        case 0x0000: di = valueData; break;
-        case 0x4000: di = directoryData; break;
+        case 0x0000: return valueData;
+        case 0x4000: return directoryData;
+        default: throw Error(kerCorruptedMetadata);
         }
-        return di;
     } // CiffComponent::dataLocation
 
     /*!
diff --git a/src/crwimage_int.hpp b/src/crwimage_int.hpp
index 3e45bb73..84997d29 100644
--- a/src/crwimage_int.hpp
+++ b/src/crwimage_int.hpp
@@ -74,7 +74,6 @@ namespace Exiv2 {
 
     //! Type to identify where the data is stored in a directory
     enum DataLocId {
-        invalidDataLocId,
         valueData,
         directoryData,
         lastDataLocId
diff --git a/tests/bugfixes/github/test_issue_841.py b/tests/bugfixes/github/test_issue_841.py
new file mode 100644
index 00000000..14ba1430
--- /dev/null
+++ b/tests/bugfixes/github/test_issue_841.py
@@ -0,0 +1,22 @@
+# -*- coding: utf-8 -*-
+
+from system_tests import CaseMeta, path, check_no_ASAN_UBSAN_errors
+
+
+class InvalidDataLocationInCiffComponentDoRead(metaclass=CaseMeta):
+    """
+    Regression test for the bug described in:
+    https://github.com/Exiv2/exiv2/issues/841
+
+    An invalid data location causes an assertion failure.
+    """
+    url = "https://github.com/Exiv2/exiv2/issues/841"
+
+    filename = path("$data_path/issue_841_poc.crw")
+    commands = ["$exiv2 $filename"]
+    stdout = [""]
+    stderr = [
+        """$exiv2_exception_message $filename:
+$kerCorruptedMetadata
+"""]
+    retval = [1]
-- 
2.22.0

